lib/cache.gi

```diff
@@ -0,0 +1,44 @@
+#############################################################################
+##
+#W  cache.gi                     GAP library                 Chris Jefferson
+##
+##
+#Y  Copyright (C) 2017 University of St Andrews, Scotland
+##
+##  This file defines various types of caching data structures
+##
+##  Note that this file is read very early in GAP's startup, so we cannot
+##  make this a method and use method dispatch, or other bits of nice
+##  GAP functionality.
+##
+
+InstallGlobalFunction(MemoizePosIntFunction,
+function(func, defaults...)
+    local boundvals, original, uniqueobj;
+
+    # This is an object which cannot exist anywhere else
+    uniqueobj := "";
+
+    if LEN_LIST(defaults) = 0 then
+        original := AtomicList([]);
+    else
+        original := AtomicList(defaults[1]);
+    fi;
+
+    boundvals := MakeWriteOnceAtomic(AtomicList(original));
+
+    InstallMethod(FlushCaches, [],
+        function()
+            boundvals := MakeWriteOnceAtomic(AtomicList(original));
+        end);
+
+    return function(val)
+        local v;
+        v := GetWithDefault(boundvals, val, uniqueobj);
+        if IsIdenticalObj(v, uniqueobj) then
+            v := func(val);
+            boundvals[val] := v;
```

Comment:
Isn't this still suspect to a race? If initially `boundvals[val]` is bound, we compute our `v`. We then store it -- but another thread may have raced us, so you store is silently rejected. But we still return `v`, which now may be different from the (hopefully equivalent, but still non-identical) value seen by other threads.

Which leads me back to my original suggestions:
```
if IsIdenticalObj(v, uniqueobj) then
  boundvals[val] := func(val);
  v := boundvals[val];
fi;
```